File: 9655 돌 게임/Main.java
// 9655 돌 게임
import java.io.*;
import static java.lang.Math.min;

public class Main{
	
	private static final int[] pick = {1, 3};
	private static final int SK = 1, CY = 2;
	private int stone;
	private int[][] dp;
	
	public static void main(String[] args){
		(new Main()).solve();
	}

	private void solve(){
		this.inputStone();
		this.initDp();
		int pickResult = this.pickStone(this.stone, this.SK);
		System.out.println((pickResult == this.SK ? "SK" : "CY"));
	}
	
	private void inputStone(){
		try(BufferedReader br = new BufferedReader(new InputStreamReader(System.in))){
			this.stone = Integer.parseInt(br.readLine());
		}catch(IOException IOE){}
	}
	
	private void initDp(){
		this.dp = new int[this.stone+1][3];
	}
	
	private int pickStone(int remainStone, int turn){
		if(turn == 0) return nextTurn(turn);
		if(this.dp[remainStone][turn] != 0) return this.dp[remainStone][turn];
		for(int i = 0; i < this.pick.length; i++){
			if(remainStone - this.pick[i] < 0) continue;
			int winPlayer = this.pickStone(remainStone-this.pick[i], this.nextTurn(turn));
			if(winPlayer == turn) this.dp[remainStone][turn] = winPlayer;
		}
		if(this.dp[remainStone][turn] == 0) this.dp[remainStone][turn] = this.nextTurn(turn);
		return dp[remainStone][turn];
	}
	
	private int nextTurn(int nowTurn){
		return nowTurn == this.SK ? this.CY : this.SK;
	}
	
}

File: 9655 돌 게임/Main.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.math.abs

val picked = listOf(1, 3)
val dp = mutableListOf<MutableList<Int>>()

fun main() {
    init()
    val line = BufferedReader(InputStreamReader(System.`in`)).readLine()
    println(
        when (play(0, line.toInt())) {
            0 -> "SK"
            else -> "CY"
        }
    )
}

private fun init() {
    for (i in 0..1000) {
        dp.add(mutableListOf(-1, -1))
    }
}

private fun play(turn: Int, leftStones: Int): Int {
    if (dp[leftStones][turn] > -1) {
        return dp[leftStones][turn]
    }
    if (leftStones == 0) {
        return when (turn) {
            1 -> 0
            else -> 1
        }
    }
    var canWin = false
    for (pick in picked) {
        if (leftStones - pick < 0) {
            continue
        }
        dp[leftStones - pick][turn.nextTurn()] = play(turn.nextTurn(), leftStones - pick)
        canWin = when (dp[leftStones - pick][turn.nextTurn()]) {
            turn -> true
            else -> false
        }
    }
    return when (canWin) {
        true -> turn
        else -> turn.nextTurn()
    }
}

private fun Int.nextTurn(): Int = abs(this - 1)
